API Scaffolders can now be more expressive when registering fields

ScopeyNZ opened the pull request:
This adds the ability to register a new `FieldDefinition` when adding fields with a scaffolder. This means you can explicitly specify the type AND the resolver for the field rather than relying on it's resolution through the field type of the field on the dataobject.

Part of the reasoning for this PR is it allows scaffolders to define types (ie. boolean, text, int) rather than using `->obj` which falls back to "text" for everything that has no casting (or `$db`) defined. Eg:

```
$scaffolder->addField('MethodAccess', Type::bool(), function($obj) { 
     return $obj->myMethod();
}
```

This PR also tidies up how the fields are retained within the scaffolder, rather than being a non-associative array and then making sure the array is distinct on the field name, I updated it to just be associative by field name. Slight performance gain I suspect.

Fixes #162

ScopeyNZ commented:
I feel like `ArrayData` is the wrong way to store field information. To have objects in an `ArrayData` it needs to extend `ViewableData` but why should a `FieldDefinition` extend `ViewableData`? 

I'm pretty keen to work on making this more intuitive. I might be able to come up with a BC way to implement this before 4.3 but for now this is ready to go. Hopefully it's acceptable as is, although it feels slightly hacky with the `$field->hasField('Definition')` and `$map = $field->toMap(); $map['Definition']` stuff.

ScopeyNZ commented:
Sounds fair. I'll try and get some time to address this before Friday.

unclecheese commented:
The `fields` array is already quite overloaded. You might want to break it out into something like `casting`, because there's a reasonable case for overloading the magic casting for native `db` fields, e.g. an `Int` casted as `Bool` or vice versa.

maxime-rainville commented on src/Scaffolding/Scaffolders/DataObjectScaffolder.php:
```diff
@@ -676,37 +707,50 @@ protected function createFields(Manager $manager)
-                        $this->getDataObjectClass()
-                    )
-                );
+        foreach ($this->fields as $fieldName => $data) {
```
This method is getting kind of big and difficult to follow. Can we split it up into smaller private methods?

maxime-rainville commented on src/Scaffolding/StaticSchema.php:
```diff
@@ -135,6 +135,27 @@ public function isValidFieldName(ViewableData $instance, $fieldName)
+     * @param ViewableData $instance
+     * @param $fieldName
+     */
+    public function assertValidFieldName(ViewableData $instance, $fieldName)
```
The entire point of this method is to throw an exception. We should include a `@throws` statement.

maxime-rainville commented on src/FieldDefinition.php:
```diff
@@ -0,0 +1,70 @@
+use GraphQL\Type\Definition\Type;
+use SilverStripe\GraphQL\Scaffolding\Extensions\TypeCreatorExtension;
+
+class FieldDefinition
```
There's no unit test for this class.

maxime-rainville commented on src/Scaffolding/Scaffolders/DataObjectScaffolder.php:
```diff
@@ -102,27 +104,38 @@ public function addFields(array $fieldData)
-                [
+
+            // If we're given a field definition we can accept that interface
+            if ($data instanceof FieldDefinition) {
```
Can we update the unit test in `tests/Scaffolding/Scaffolders/DataObjectScaffolderTest.php` to test `FieldDefinition`?

maxime-rainville commented on src/Scaffolding/Scaffolders/DataObjectScaffolder.php:
```diff
@@ -102,27 +104,38 @@ public function addFields(array $fieldData)
      * @return $this
      */
-    public function addField($field, $description = null)
+    public function addField($field, $definition = null)
```
Can we update the unit test in `tests/Scaffolding/Scaffolders/DataObjectScaffolderTest.php` to test `FieldDefinition`?

maxime-rainville commented on src/Scaffolding/Scaffolders/DataObjectScaffolder.php:
```diff
@@ -645,6 +663,20 @@ protected function nestedConnections()
         return $queries;
     }
 
+    protected function assertValidFieldData($data, $fieldName)
```
Can we document this method?

unclecheese commented on src/Scaffolding/Scaffolders/DataObjectScaffolder.php:
```diff
@@ -102,27 +104,38 @@ public function addFields(array $fieldData)
             $name = $assoc ? $k : $data;
-            $field = ArrayData::create(
-                [
+
```
Let's use `ArrayLib::is_assoc()` for consistency.

unclecheese commented on src/Scaffolding/Scaffolders/DataObjectScaffolder.php:
```diff
@@ -222,14 +237,11 @@ public function setFieldDescription($field, $description)
-                    'Description' => $description
-                ]
-            )
+        $this->fields[$field] = ArrayData::create(
```
I think this should be its own subclass of `ArrayData`. That way you can do your own validation, getters/setters, etc, and everything is more expressive.

unclecheese commented on src/Scaffolding/Scaffolders/DataObjectScaffolder.php:
```diff
@@ -408,7 +420,13 @@ public function cloneTo(DataObjectScaffolder $target)
-                $target->addField($field->Name, $field->Description);
+                $arg = $field->Description;
+                // Work around `ArrayData` not supporting objects as values (unless they extend "ViewableData")
+                if ($field->hasField('Definition')) {
```
Seems to me you wouldn't have to do this hack if you created your own `ArrayData` subclass.

ScopeyNZ replied:
I agree. I don't even think it should be a subclass of ArrayData as it's quite opinionated around holding objects that extend `ViewableData`.

ScopeyNZ commented:
Thanks for the reviews guys! I've blocked a couple of hours out tomorrow afternoon to run over this some more so I'll try to address everything then.

ScopeyNZ commented:
I've added a commit - some details on the change are in the commit message.

I refactored this PR so that the only way that fields are maintained within `DataObjectScaffolder` is by using an associative array of `FieldDefinition`s. It occurred to me that this new object is only relevant for `DataObjectScaffolder` so I moved it a more relevant namespace.

I still need to increase the test coverage which I'll try to do before next week but I appreciate any comments you guys may have in the meantime.

ScopeyNZ commented:
Also this is definitely breaking changes now :( - And isn't actually a dependency for the CCs elemental work in retrospect - We can drop it from the 4.3 release milestone (although it'd be awesome to get this into the new 3.0 release).

ScopeyNZ commented:
I guess it probably does. This has breaking changes though and it's not high on my list to address this right now.

ScopeyNZ commented:
I don't have access to this repo for some reason so I can't link PR (nor edit the issue to mention a PR)

robbieaverill commented:
>I don't have access to this repo for some reason so I can't link PR (nor edit the issue to mention a PR)

Core team should, I've fixed that

maxime-rainville commented:
@ScopeyNZ Do you still have work to do on this? We've got it in the _blocked_ column on the OS board.

ScopeyNZ commented:
I can't remember where I left it. Looks like there's at least a failed test to resolve. I'll put aside some time this week to look over it.

unclecheese commented:
I hate to take this conversation in a totally different direction, but I'm wondering if, since it appears version 4 will use generated code from a dev task, maybe we should rely on statically analysed type hints for this?

Option A: Merge this into 3 only, with a new static analysis implementation in 4
Option B: Merge this into 3 and deprecate it in 4, supporting both for one major release cycle
Option C: Don't merge this. Let's just wait until 4, when we can do it right.
Option D: Merge this and support both long term (because maybe there's a legit case for having a GraphQL type that differs from your PHP type?)

ScopeyNZ commented:
Considering that the API is changing so quickly and we're jumping to new majors so often I don't think it's worth increasing maintenance burden in the `3` branch. (Option C).

I still haven't found the time to run over this PR - it's not currently relevant for me.

maxime-rainville commented:
Let's close for now. We can reopen later if we find it's relevant.
